ai_server/server.py

```diff
@@ -1,12 +1,250 @@
+        raise Exception(f"Llama-server request timed out for model {model}")
+    except requests.RequestException as e:
+        raise Exception(f"Llama-server request failed: {str(e)}")
+
+def get_available_llamacpp_models():
+    """Return list of available llama.cpp models."""
+    models = []
+    for model_dir in LLAMA_CPP_MODEL_DIRS:
+        dir_path = os.path.join(GGUF_DIR, model_dir)
+        if os.path.exists(dir_path):
+            try:
+                files = os.listdir(dir_path)
+                for file in files:
+                    if file.endswith('.gguf'):
+                        models.append(f"{model_dir}/{file}")
+            except Exception:
+                continue
+    return models
+
+def resolve_model_path(model: str) -> Optional[str]:
+    """Resolve model name to full GGUF file path."""
+    for model_dir in LLAMA_CPP_MODEL_DIRS:
+        dir_path = os.path.join(GGUF_DIR, model_dir)
+        if not os.path.exists(dir_path):
+            continue
+            
+        try:
+            files = os.listdir(dir_path)
+            gguf_files = [f for f in files if f.endswith('.gguf')]
+            
+            for file in gguf_files:
+                if (model == model_dir or  # Directory name match
+                    model == f"{model_dir}/{file}" or  # Full path match
+                    model == file or  # Filename match
+                    model == os.path.splitext(file)[0]):  # Name without extension
+                    return os.path.join(dir_path, file)
+        except Exception:
+            continue
+    
+    return None
+
+def is_llamacpp_available(model: str) -> bool:
+    """Check if model is available in llama.cpp."""
+    return resolve_model_path(model) is not None
+
+def chat_with_ollama(model: str, content: str) -> str:
+    """Handle chat using ollama."""
+    response = ollama.chat(
+        model=model,
+        messages=[{'role': 'user', 'content': content}],
+        stream=False
+    )
+    return response.message.content
+
+def chat_with_llamacpp(model: str, content: str, timeout: int = 300) -> str:
+    """Handle chat using llama.cpp CLI."""
+    model_path = resolve_model_path(model)
+    
+    if not model_path:
+        available_models = get_available_llamacpp_models()
+        raise ValueError(f"Model not found: {model}. Available models: {available_models}")
+    
+    cmd = [
+        LLAMA_CPP_CLI,
+        '-m', model_path,
+        '--n-gpu-layers', '40',
+        '-p', content,
+        '-n', '512',
+        '--no-display-prompt',
+        '-no-cnv'
+    ]
+    
+    try:
+        result = subprocess.run(
+            cmd,
+            capture_output=True,
+            text=False,
+            timeout=timeout,
+            check=True
+        )
+        
+        # For V3 IQ1_M and IQ2_XXS models
+        try:
+            stdout_text = result.stdout.decode('utf-8')
+        except UnicodeDecodeError:
+            stdout_text = result.stdout.decode('utf-8', errors='replace')
```

Comment:
This is fine but I don't think you need this in a try-except, just always use the `errors='replace'` argument.